pkg/splunk/enterprise/util.go

```diff
@@ -292,3 +343,47 @@ func DeleteOwnerReferencesForS3SecretObjects(client splcommon.ControllerClient,
 
 	return err
 }
+
+// GetAppListFromS3Bucket gets the list of apps from remote storage.
+func GetAppListFromS3Bucket(client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterprisev1.AppFrameworkSpec) (map[string]*splclient.S3Response, error) {
+
+	scopedLog := log.WithName("GetAppListFromS3Bucket").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	sourceToAppListMap := make(map[string]*splclient.S3Response)
+
+	volMap := make(map[string]enterprisev1.VolumeSpec)
+
+	for _, vol := range appFrameworkRef.VolList {
+		volMap[vol.Name] = vol
+	}
+
+	scopedLog.Info("Getting the list of apps from remote storage...")
+
+	var s3Response *splclient.S3Response
+	var vol enterprisev1.VolumeSpec
+
+	for _, appSource := range appFrameworkRef.AppSources {
+		// get the volume spec from the volume name
+		if appSource.AppSourceDefaultSpec.VolName != "" {
+			vol = volMap[appSource.AppSourceDefaultSpec.VolName]
+		} else {
+			vol = volMap[appFrameworkRef.Defaults.VolName]
+		}
+
+		//get the corresponding S3 client
+		s3Client, err := GetRemoteStorageClient(client, cr, appFrameworkRef, &vol, appSource.Location)
+		if err != nil {
+			return nil, err
```

Comment:
Needs further thought when we actually start supporting multiple sources per CR - do we really want to quit right after failure of one source?

Reply:
Hmm good point. We are actually supporting multiple appSources, so this might bite us. We can just log an error and continue with the next appSource. Thoughts @sgontla and @jryb ?